File: Backend/twetter/src/main/java/com/bartdebever/twetter/services/Local/LocalCrudService.java
package com.bartdebever.twetter.services.Local;

import com.bartdebever.twetter.helpers.CSharp;
import com.bartdebever.twetter.models.interfaces.IEntity;
import com.bartdebever.twetter.services.interfaces.ICrudService;

import java.util.ArrayList;
import java.util.List;

public abstract class LocalCrudService <T extends IEntity> implements ICrudService<T> {
    private List<T> entityList;

    LocalCrudService() {
        entityList = new ArrayList<>();
    }

    @Override
    public void update(T entity) {
        checkForNull(entity);
        T existingEntity = CSharp.firstOrDefault(entityList, entity.getId());
        entityList.set(entityList.indexOf(existingEntity), entity);
    }

    @Override
    public void delete(T entity) {
        checkForNull(entity);
        T existingEntity = CSharp.firstOrDefault(entityList, entity.getId());
        entityList.remove(existingEntity);
    }

    @Override
    public void insert(T entity) {
        checkForNull(entity);
        entityList.add(entity);
    }

    @Override
    public List<T> getAll() {
        return entityList;
    }

    @Override
    public T getById(int id) {
        if (id <= 0) {
            throw new IllegalArgumentException("Id must be greater than 0");
        }

        return CSharp.firstOrDefault(entityList, id);
    }

    private void checkForNull(T entity) throws IllegalArgumentException {
        if (entity == null) {
            throw new IllegalArgumentException("entity can not be null");
        }
    }

    List<T> getEntityList() {
        return entityList;
    }
}

File: Backend/src/main/kotlin/services/Local/LocalCrudService.kt
package services.Local

import helpers.CSharp
import interfaces.IEntity
import services.interfaces.ICrudService

import java.util.ArrayList

abstract class LocalCrudService<T : IEntity> internal constructor() : ICrudService<T> {
    val entityList: MutableList<T>

    override val all: List<T>
        get() = entityList

    init {
        entityList = ArrayList()
    }

    override fun update(entity: T) {
        checkForNull(entity)
        val existingEntity = entityList.firstOrNull { x -> x.id == entity.id }
        entityList[entityList.indexOf(existingEntity)] = entity
    }

    override fun delete(entity: T) {
        checkForNull(entity)
        val existingEntity = entityList.firstOrNull { x -> x.id == entity.id }
        entityList.remove(existingEntity)
    }

    override fun insert(entity: T) {
        checkForNull(entity)
        entityList.add(entity)
    }

    override fun getById(id: Int): T? {
        if (id <= 0) {
            throw IllegalArgumentException("Id must be greater than 0")
        }

        return entityList.firstOrNull { x -> x.id == id }
    }

    @Throws(IllegalArgumentException::class)
    private fun checkForNull(entity: T?) {
        if (entity == null) {
            throw IllegalArgumentException("entity can not be null")
        }
    }

    internal fun getEntityList(): List<T> {
        return entityList
    }
}
